### PR TITLE
[DataPipe] Adding LengthSetterIterDataPipe

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #747

This DataPipe allows users to manually set length of an `IterDataPipe` with no other side effect. This is useful for DataPipes whose final length cannot be known in advance (e.g. ``filter``). If you know the final length with certainty, you can manually set it for usages by DataLoader or other DataPipes.

Previously, users theoretically could use `.header(length)` to manually set length. However, the UX is suboptimal in that 1) the name is not obvious, and 2) warnings are raised unless the HeaderDataPipe has been fully traversed once to confirm the length.

Differential Revision: [D38989359](https://our.internmc.facebook.com/intern/diff/D38989359)

### CODE REVIEW

File: torchdata/datapipes/iter/util/header.py [hunk 1]
```diff
@@ -60,3 +61,37 @@ def __len__(self) -> int:
                 "The actual value may be smaller if the actual length of source_datapipe is smaller than the limit."
             )
             return self.limit
+
+
+@functional_datapipe("set_length")
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/header.py:66]: What would be the difference between adding this DataPipe and implement `set_length` API to the base class of `IterDataPipe`?

**Comment by NivekT** [comment]
Per @ejguan 's request, I am adding this to core as an alternative to this PR: https://github.com/pytorch/pytorch/pull/83750

I will close this one if we decide the other PR is better.

**Comment by NivekT** [comment]
We will be landing this PR instead of https://github.com/pytorch/pytorch/pull/83750.

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38989359).

File: torchdata/datapipes/iter/util/header.py [hunk 2]
```diff
@@ -60,3 +61,44 @@ def __len__(self) -> int:
...
+        >>> header_dp = IterableWrapper(range(10)).filter(lambda x: x < 5).header(3)
+        >>> list(header_dp)  # Use `.header()` if you want to limit the number of elements yielded
+        [0, 1, 2]
+        >>> len(header_dp)
+        3
+    """
+
+    def __init__(self, source_datapipe: IterDataPipe[T_co], length: int) -> None:
+        self.source_datapipe: IterDataPipe[T_co] = source_datapipe
+        self.length: int = length
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/header.py:98]: Could we add one validation on the `length >= 0`.

**Review by ejguan — APPROVED** [review]
One nit comment. Otherwise, LGTM. Thank you

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38989359).

**Pull request closed without merging**